# 五、数据处理 
### 1.数据清洗 
#### 数据分析的第一步是提高数据质量。数据清洗要做的就是处理缺失数据以及清除无意义的信息。这是数据价值链中最关键的步骤。垃圾数 据，即使是通过最好的分析，也将产生错误的结果，并误导业务本身。



#### 1）重复值的处理
#### drop_duplicates()    把数据结构中行相同的数据去除(保留其中的一行)

In [1]:
from pandas import DataFrame
from pandas import read_excel
df = read_excel('data/rz2.xlsx')
df

YHM          TCSJ          YWXT                 IP  \
0   S1402048  1.892225e+10  1.225790e+17      221.205.98.55   
1   S1411023  1.352226e+10  1.225790e+17    183.184.226.205   
2   S1402048  1.342226e+10           NaN      221.205.98.55   
3   20031509  1.882226e+10           NaN      222.31.51.200   
4   S1405010  1.892225e+10  1.225790e+17       120.207.64.3   
5   20140007           NaN  1.225790e+17      222.31.51.200   
6   S1404095  1.382225e+10  1.225790e+17      222.31.59.220   
7   S1402048  1.332225e+10  1.225790e+17  221.205.98.55       
8   S1405011  1.892226e+10  1.225790e+17     183.184.230.38   
9   S1402048  1.332225e+10  1.225790e+17  221.205.98.55       
10  S1405011  1.892226e+10  1.225790e+17     183.184.230.38   

                   DLSJ  
0   2014-11-04 08:44:46  
1   2014-11-04 08:45:06  
2   2014-11-04 08:46:39  
3   2014-11-04 08:47:41  
4   2014-11-04 08:49:03  
5   2014-11-04 08:50:06  
6   2014-11-04 08:50:02  
7   2014-11-04 08:49:18  
8   2014-11-04 08:14:55  
9   2014-11-04 08:49:18  
10  2014-11-04 08:14:55

In [2]:
newDF=df.drop_duplicates()
newDF

YHM          TCSJ          YWXT                 IP  \
0  S1402048  1.892225e+10  1.225790e+17      221.205.98.55   
1  S1411023  1.352226e+10  1.225790e+17    183.184.226.205   
2  S1402048  1.342226e+10           NaN      221.205.98.55   
3  20031509  1.882226e+10           NaN      222.31.51.200   
4  S1405010  1.892225e+10  1.225790e+17       120.207.64.3   
5  20140007           NaN  1.225790e+17      222.31.51.200   
6  S1404095  1.382225e+10  1.225790e+17      222.31.59.220   
7  S1402048  1.332225e+10  1.225790e+17  221.205.98.55       
8  S1405011  1.892226e+10  1.225790e+17     183.184.230.38   

                  DLSJ  
0  2014-11-04 08:44:46  
1  2014-11-04 08:45:06  
2  2014-11-04 08:46:39  
3  2014-11-04 08:47:41  
4  2014-11-04 08:49:03  
5  2014-11-04 08:50:06  
6  2014-11-04 08:50:02  
7  2014-11-04 08:49:18  
8  2014-11-04 08:14:55

#### 2) 缺失值处理
#### 对于缺失数据的处理方式有数据补齐、删除对应行、不处理等方法。
#### (1)	dropna()  去除数据结构中值为空的数据行

In [4]:
from pandas import DataFrame
from pandas import read_excel
df = read_excel('data/rz2.xlsx')
newDF=df.dropna()
newDF

YHM          TCSJ          YWXT                 IP  \
0   S1402048  1.892225e+10  1.225790e+17      221.205.98.55   
1   S1411023  1.352226e+10  1.225790e+17    183.184.226.205   
4   S1405010  1.892225e+10  1.225790e+17       120.207.64.3   
6   S1404095  1.382225e+10  1.225790e+17      222.31.59.220   
7   S1402048  1.332225e+10  1.225790e+17  221.205.98.55       
8   S1405011  1.892226e+10  1.225790e+17     183.184.230.38   
9   S1402048  1.332225e+10  1.225790e+17  221.205.98.55       
10  S1405011  1.892226e+10  1.225790e+17     183.184.230.38   

                   DLSJ  
0   2014-11-04 08:44:46  
1   2014-11-04 08:45:06  
4   2014-11-04 08:49:03  
6   2014-11-04 08:50:02  
7   2014-11-04 08:49:18  
8   2014-11-04 08:14:55  
9   2014-11-04 08:49:18  
10  2014-11-04 08:14:55

####  (2)	df.fillna()  用其他数值替代NaN
#### 有些时候空数据直接删除会影响分析的结果，可以对数据进行填补。 

In [148]:
from pandas import DataFrame
from pandas import read_excel
df = read_excel('data/rz2.xlsx')
df.fillna('?')

YHM         TCSJ         YWXT                 IP                 DLSJ
0   S1402048  1.89223e+10  1.22579e+17      221.205.98.55  2014-11-04 08:44:46
1   S1411023  1.35223e+10  1.22579e+17    183.184.226.205  2014-11-04 08:45:06
2   S1402048  1.34223e+10            ?      221.205.98.55  2014-11-04 08:46:39
3   20031509  1.88223e+10            ?      222.31.51.200  2014-11-04 08:47:41
4   S1405010  1.89223e+10  1.22579e+17       120.207.64.3  2014-11-04 08:49:03
5   20140007            ?  1.22579e+17      222.31.51.200  2014-11-04 08:50:06
6   S1404095  1.38223e+10  1.22579e+17      222.31.59.220  2014-11-04 08:50:02
7   S1402048  1.33223e+10  1.22579e+17  221.205.98.55      2014-11-04 08:49:18
8   S1405011  1.89223e+10  1.22579e+17     183.184.230.38  2014-11-04 08:14:55
9   S1402048  1.33223e+10  1.22579e+17  221.205.98.55      2014-11-04 08:49:18
10  S1405011  1.89223e+10  1.22579e+17     183.184.230.38  2014-11-04 08:14:55

In [149]:
from pandas import DataFrame
from pandas import read_excel
df = read_excel('data/rz2_0.xlsx')
df

No  math  physical  Chinese
0   1  76.0        85     78.0
1   2  85.0        56      NaN
2   3  76.0        95     85.0
3   4   NaN        75     58.0
4   5  87.0        52     68.0

In [150]:
df.fillna(method='pad')  # 用前一个数据值替代NaN

No  math  physical  Chinese
0   1  76.0        85     78.0
1   2  85.0        56     78.0
2   3  76.0        95     85.0
3   4  76.0        75     58.0
4   5  87.0        52     68.0

In [151]:
df.fillna(method='bfill')  #用后一个数据值替代NaN

No  math  physical  Chinese
0   1  76.0        85     78.0
1   2  85.0        56     85.0
2   3  76.0        95     85.0
3   4  87.0        75     58.0
4   5  87.0        52     68.0

In [152]:
from pandas import DataFrame
from pandas import read_excel
df = read_excel('data/rz2_0.xlsx')
df

No  math  physical  Chinese
0   1  76.0        85     78.0
1   2  85.0        56      NaN
2   3  76.0        95     85.0
3   4   NaN        75     58.0
4   5  87.0        52     68.0

In [153]:
df.fillna(df.mean())  #用平均数或者其他描述性统计量来代替NaN。

No  math  physical  Chinese
0   1  76.0        85    78.00
1   2  85.0        56    72.25
2   3  76.0        95    85.00
3   4  81.0        75    58.00
4   5  87.0        52    68.00

In [35]:
#为某列使用该列的均值来填补数据
from pandas import DataFrame
from pandas import read_excel
df = read_excel('data/rz2_0.xlsx')
df.fillna(df.mean()['math':'Chinese'])#各自列的缺失值用该列的平均值替代,而且列名必须从左到右

No  math  physical  Chinese
0   1  76.0        85    78.00
1   2  85.0        56    72.25
2   3  76.0        95    85.00
3   4  81.0        75    58.00
4   5  87.0        52    68.00

In [36]:
#strip()：清除字符型数据左右(首尾)指定的字符，默认为空格，中间的不清除
from pandas import DataFrame
from pandas import read_excel
df = read_excel('data/rz2.xlsx')
newDF=df['IP'].str.strip()    #因为IP是一个对象，所以先转为str。
newDF

0       221.205.98.55
1     183.184.226.205
2       221.205.98.55
3       222.31.51.200
4        120.207.64.3
5       222.31.51.200
6       222.31.59.220
7       221.205.98.55
8      183.184.230.38
9       221.205.98.55
10     183.184.230.38
Name: IP, dtype: object

In [45]:
df.IP[2],newDF[2]

('   221.205.98.55', '221.205.98.55')

### 2.数据抽取 
#### (1) 字段抽取：抽出某列上指定位置的数据做成新的列。
> slice(start,stop)
> start   开始位置
> stop   结束位置

In [154]:
from pandas import DataFrame
from pandas import read_excel
df = read_excel('data/rz2.xlsx')
df

YHM          TCSJ          YWXT                 IP  \
0   S1402048  1.892225e+10  1.225790e+17      221.205.98.55   
1   S1411023  1.352226e+10  1.225790e+17    183.184.226.205   
2   S1402048  1.342226e+10           NaN      221.205.98.55   
3   20031509  1.882226e+10           NaN      222.31.51.200   
4   S1405010  1.892225e+10  1.225790e+17       120.207.64.3   
5   20140007           NaN  1.225790e+17      222.31.51.200   
6   S1404095  1.382225e+10  1.225790e+17      222.31.59.220   
7   S1402048  1.332225e+10  1.225790e+17  221.205.98.55       
8   S1405011  1.892226e+10  1.225790e+17     183.184.230.38   
9   S1402048  1.332225e+10  1.225790e+17  221.205.98.55       
10  S1405011  1.892226e+10  1.225790e+17     183.184.230.38   

                   DLSJ  
0   2014-11-04 08:44:46  
1   2014-11-04 08:45:06  
2   2014-11-04 08:46:39  
3   2014-11-04 08:47:41  
4   2014-11-04 08:49:03  
5   2014-11-04 08:50:06  
6   2014-11-04 08:50:02  
7   2014-11-04 08:49:18  
8   2014-11-04 08:14:55  
9   2014-11-04 08:49:18  
10  2014-11-04 08:14:55

In [155]:
df['TCSJ'].dtype

dtype('float64')

In [156]:
df['TCSJ']=df['TCSJ'].astype(np.str)   #astype()转化类型
df['TCSJ']

0     18922254812.0
1     13522255003.0
2     13422259938.0
3     18822256753.0
4     18922253721.0
5               nan
6     13822254373.0
7     13322252452.0
8     18922257681.0
9     13322252452.0
10    18922257681.0
Name: TCSJ, dtype: object

In [157]:
bands = df['TCSJ'].str.slice(0,3)
bands

0     189
1     135
2     134
3     188
4     189
5     nan
6     138
7     133
8     189
9     133
10    189
Name: TCSJ, dtype: object

In [160]:
areas= df['TCSJ'].str.slice(3,7);
areas

0     2225
1     2225
2     2225
3     2225
4     2225
5         
6     2225
7     2225
8     2225
9     2225
10    2225
Name: TCSJ, dtype: object

In [19]:
tell= df['TCSJ'].str.slice(7,11);
tell

0     4812
1     5003
2     9938
3     6753
4     3721
5         
6     4373
7     2452
8     7681
9     2452
10    7681
Name: TCSJ, dtype: object

#### (2) 字段拆分：按指定的字符sep，拆分已有的字符串。
> split(sep,n,expand=False)   
		  > sep :用于分隔字符串的分隔符。
		  > n :分割后新增的列数。
		  > expand: 是否展开为数据框，默认为False。
>#### 返回值：expand为True，返回DaraFrame；False返回Series。 

In [93]:
from pandas import DataFrame
from pandas import read_excel
df = read_excel('data/rz2.xlsx')
newDF=df['IP'].str.strip()  #IP先转为str，再删除首位空格
newDF= df['IP'].str.split('.',1,True)#按第一个"."分成两列，1表示新增的列数
newDF

0              1
0       221      205.98.55
1       183    184.226.205
2       221      205.98.55
3       222      31.51.200
4       120       207.64.3
5       222      31.51.200
6       222      31.59.220
7       221  205.98.55    
8       183     184.230.38
9       221  205.98.55    
10      183     184.230.38

In [24]:
newDF.columns = ['IP1','IP2-4']
#给第一第二列增加列名称
newDF

IP1          IP2-4
0       221      205.98.55
1       183    184.226.205
2       221      205.98.55
3       222      31.51.200
4       120       207.64.3
5       222      31.51.200
6       222      31.59.220
7       221  205.98.55    
8       183     184.230.38
9       221  205.98.55    
10      183     184.230.38

#### (3) 记录抽取：是指根据一定的条件，对数据进行抽取。
     dataframe[condition]
    	       condition：过滤条件
     返回值：DataFrame
#### 常用的condition类型：
	-> 比较运算：<、>、>=、<=、!=，如：df[df.comments>10000)]；
	-> 范围运算：between(left,right)，如：df[df.comments.between(1000,10000)]；
	-> 空置运算：pandas.isnull(column) ，如：df[df.title.isnull()];
	-> 字符匹配：str.contains(patten,na = False) ，如：
         df[df.title.str.contains(‘电台’,na=False)]
    -> 	逻辑运算：&(与)，|(或)，not(取反)；如：
      df[(df.comments>=1000)&(df.comments<=10000)] 与df[df.comments.between(1000,10000)]等价。

In [25]:
import pandas
from pandas import read_excel
df = read_excel('data/rz2.xlsx')
df[df.TCSJ==13322252452]

YHM          TCSJ          YWXT                 IP  \
7  S1402048  1.332225e+10  1.225790e+17  221.205.98.55       
9  S1402048  1.332225e+10  1.225790e+17  221.205.98.55       

                  DLSJ  
7  2014-11-04 08:49:18  
9  2014-11-04 08:49:18

In [26]:
df[df.TCSJ>13500000000]

YHM          TCSJ          YWXT                 IP  \
0   S1402048  1.892225e+10  1.225790e+17      221.205.98.55   
1   S1411023  1.352226e+10  1.225790e+17    183.184.226.205   
3   20031509  1.882226e+10           NaN      222.31.51.200   
4   S1405010  1.892225e+10  1.225790e+17       120.207.64.3   
6   S1404095  1.382225e+10  1.225790e+17      222.31.59.220   
8   S1405011  1.892226e+10  1.225790e+17     183.184.230.38   
10  S1405011  1.892226e+10  1.225790e+17     183.184.230.38   

                   DLSJ  
0   2014-11-04 08:44:46  
1   2014-11-04 08:45:06  
3   2014-11-04 08:47:41  
4   2014-11-04 08:49:03  
6   2014-11-04 08:50:02  
8   2014-11-04 08:14:55  
10  2014-11-04 08:14:55

In [27]:
df[df.TCSJ.between(13400000000,13999999999)]

YHM          TCSJ          YWXT                 IP  \
1  S1411023  1.352226e+10  1.225790e+17    183.184.226.205   
2  S1402048  1.342226e+10           NaN      221.205.98.55   
6  S1404095  1.382225e+10  1.225790e+17      222.31.59.220   

                  DLSJ  
1  2014-11-04 08:45:06  
2  2014-11-04 08:46:39  
6  2014-11-04 08:50:02

In [28]:
df[df.YWXT.isnull()]

YHM          TCSJ  YWXT                IP                 DLSJ
2  S1402048  1.342226e+10   NaN     221.205.98.55  2014-11-04 08:46:39
3  20031509  1.882226e+10   NaN     222.31.51.200  2014-11-04 08:47:41

In [29]:
df[df['YWXT'].isnull()]

YHM          TCSJ  YWXT                IP                 DLSJ
2  S1402048  1.342226e+10   NaN     221.205.98.55  2014-11-04 08:46:39
3  20031509  1.882226e+10   NaN     222.31.51.200  2014-11-04 08:47:41

In [30]:
df[df.IP.str.contains('222.',na=False)]

YHM          TCSJ          YWXT                 IP  \
3  20031509  1.882226e+10           NaN      222.31.51.200   
5  20140007           NaN  1.225790e+17      222.31.51.200   
6  S1404095  1.382225e+10  1.225790e+17      222.31.59.220   

                  DLSJ  
3  2014-11-04 08:47:41  
5  2014-11-04 08:50:06  
6  2014-11-04 08:50:02

#### (4) 随机抽样：是指随机从数据中按照一定的行数或者比例抽取数据。
    随机抽样函数：
    numpy.random.randint(start,end,num)
              start：范围的开始值；
              end：范围的结束值；
              num：抽样个数
    返回值：行的索引值序列

In [33]:
import pandas as pd   #这样写更简洁些
df =pd.read_excel('data/rz2.xlsx')
df

YHM          TCSJ          YWXT                 IP  \
0   S1402048  1.892225e+10  1.225790e+17      221.205.98.55   
1   S1411023  1.352226e+10  1.225790e+17    183.184.226.205   
2   S1402048  1.342226e+10           NaN      221.205.98.55   
3   20031509  1.882226e+10           NaN      222.31.51.200   
4   S1405010  1.892225e+10  1.225790e+17       120.207.64.3   
5   20140007           NaN  1.225790e+17      222.31.51.200   
6   S1404095  1.382225e+10  1.225790e+17      222.31.59.220   
7   S1402048  1.332225e+10  1.225790e+17  221.205.98.55       
8   S1405011  1.892226e+10  1.225790e+17     183.184.230.38   
9   S1402048  1.332225e+10  1.225790e+17  221.205.98.55       
10  S1405011  1.892226e+10  1.225790e+17     183.184.230.38   

                   DLSJ  
0   2014-11-04 08:44:46  
1   2014-11-04 08:45:06  
2   2014-11-04 08:46:39  
3   2014-11-04 08:47:41  
4   2014-11-04 08:49:03  
5   2014-11-04 08:50:06  
6   2014-11-04 08:50:02  
7   2014-11-04 08:49:18  
8   2014-11-04 08:14:55  
9   2014-11-04 08:49:18  
10  2014-11-04 08:14:55

In [34]:
r = numpy.random.randint(0,10,3)
r

array([2, 5, 2])

In [35]:
df.loc[r,:]  #抽取r行数据

YHM          TCSJ          YWXT                IP                 DLSJ
2  S1402048  1.342226e+10           NaN     221.205.98.55  2014-11-04 08:46:39
5  20140007           NaN  1.225790e+17     222.31.51.200  2014-11-04 08:50:06
2  S1402048  1.342226e+10           NaN     221.205.98.55  2014-11-04 08:46:39

#### 按照指定条件抽取数据：
    1）使用index标签选取数据：df.loc[行标签,列标签]

    df.loc['a':'b']           #选取ab两行数据，假设a，b为行索引
    df.loc[:,'TCSJ']         #选取TCSJ列的数据

    df.loc的第一个参数是行标签，第二个参数为列标签(可选参数，默认为所有列标签)，两个参数既可以是列表也可以是单个字符，如果两个参数都为列表则返回的是DataFrame，否则为Series。
    2）使用切片位置选取数据：df.iloc[行位置,列位置]

    df.iloc[1,1]       #选取第二行，第二列的值，返回的为单个值
    df.iloc[[0,2],:]  #选取第一行和第三行的数据
    df.iloc[0:2,:]     #选取第一行到第三行(不包含)的数据
    df.iloc[:,1]       #选取所有记录的第一列的值，返回的为一个Series
    df.iloc[1,:]       #选取第一行数据，返回的为一个Series

#### 说明：loc为location的缩写，iloc则为integer & location的缩写。
#### 更广义的切片方式是使用.ix，它自动根据给到的索引类型判断是使用位置还是标签进行切片。即：iloc为整型索引；loc为字符串索引。
#### Python默认的行序号是从0开始，我们称为行位置；但实际上0开始的行我们在计数时为第1行，也称为行号，是从1开始；有时index是被命名的，如'one','two','three','four'或'a','b','c','d'等字符串，我们称之为标签。loc索引的是行号、标签，不是行位置，如下例中df2.loc[1]索引的是第一行（行号为1），其实位置为0行；iloc索引的是位置，不能是标签或行号。 

In [37]:
import pandas as pd
index_loc = ['a','b']
index_iloc = [1,2]
data = [[1,2,3,4],[5,6,7,8]]
columns = ['one','two','three','four']
df1 = pd.DataFrame(data=data,index=index_loc,columns=columns)
df1

one  two  three  four
a    1    2      3     4
b    5    6      7     8

In [44]:
df1.loc['a']

one      1
two      2
three    3
four     4
Name: a, dtype: int64

In [46]:
df1.iloc[0]

one      1
two      2
three    3
four     4
Name: a, dtype: int64

In [38]:
df2 = pd.DataFrame(data=data,index=index_iloc,columns=columns)
df2

one  two  three  four
1    1    2      3     4
2    5    6      7     8

In [47]:
df2.iloc[0]    #或者df2.loc[1]，这里1是索引名

one      1
two      2
three    3
four     4
Name: 1, dtype: int64

#### 3）通过逻辑指针进行数据切片：df[逻辑条件]
    df[df. TCSJ >= 18822256753]   #单个逻辑条件
    df[(df. TCSJ >=13422259938 )&(df. TCSJ < 13822254373)] #多个逻辑条件组合
    这种方式获取的数据切片都是DataFrame。

In [50]:
df[df.TCSJ >= 18822256753]

YHM          TCSJ          YWXT              IP                 DLSJ
0   S1402048  1.892225e+10  1.225790e+17   221.205.98.55  2014-11-04 08:44:46
3   20031509  1.882226e+10           NaN   222.31.51.200  2014-11-04 08:47:41
4   S1405010  1.892225e+10  1.225790e+17    120.207.64.3  2014-11-04 08:49:03
8   S1405011  1.892226e+10  1.225790e+17  183.184.230.38  2014-11-04 08:14:55
10  S1405011  1.892226e+10  1.225790e+17  183.184.230.38  2014-11-04 08:14:55

#### （5） 字典数据：将字典数据抽取为dataframe，有三种方法。

In [99]:
#1.字典的key和value各作为一列
import pandas
from pandas import DataFrame
d1={'a':'[1,2,3]','b':'[0，1,2]'}
a1=pandas.DataFrame.from_dict(d1, orient='index') #将字典转化为dataframe，且key列做成了index
a1

0
a  [1,2,3]
b  [0，1,2]

In [56]:
a1.index.name = 'key'  #将index的列名改成‘key’
a1

0
key         
a    [1,2,3]
b    [0，1,2]

In [57]:
b1=a1.reset_index()    #重新增加index，并将原index做成了‘key’列
b1.columns=['key','value'] #对列重新命名为'key'和'value'
b1

key    value
0   a  [1,2,3]
1   b  [0，1,2]

In [60]:
#2.字典里的每一个元素作为一列(同长)
d2={'a':[1,2,3],'b':[4,5,6]}    #字典的value必须长度相等
a2= DataFrame(d2)
a2

a  b
0  1  4
1  2  5
2  3  6

In [66]:
#3.字典里的每一个元素作为一列(不同长)
d = {'one' : pd.Series([1, 2, 3]),'two' : pd.Series([1, 2, 3, 4])} #字典的value长度可以不相等
df = pd.DataFrame(d)
df

one  two
0  1.0    1
1  2.0    2
2  3.0    3
3  NaN    4

In [161]:
#也可以如下处理：
import pandas as pd
import numpy as np
d = dict( A = np.array([1,2]), B = np.array([1,2,3,4]))
DataFrame(dict([(k,pd.Series(v)) for k,v in d.items()]))

A  B
0  1.0  1
1  2.0  2
2  NaN  3
3  NaN  4

In [163]:
d

{'A': array([1, 2]), 'B': array([1, 2, 3, 4])}

In [105]:
#还可以处理如下： 
my_dict = dict( A = np.array([1,2]),B = np.array([1,2,3,4]) )
df = pd.DataFrame.from_dict(my_dict,orient='index')
df

0  1    2    3
A  1  2  NaN  NaN
B  1  2  3.0  4.0

In [107]:
df = pd.DataFrame.from_dict(my_dict,orient='index').T
df

A    B
0  1.0  1.0
1  2.0  2.0
2  NaN  3.0
3  NaN  4.0

In [69]:
my_dict

{'A': array([1, 2]), 'B': array([1, 2, 3, 4])}

## 3.插入记录 

In [130]:
import pandas as pd
df=pd.DataFrame({'a':[1,2,3],'b':['a','b','c'],'c':['A','B','C']})
df

a  b  c
0  1  a  A
1  2  b  B
2  3  c  C

In [131]:
df.columns[0],df.columns[1],df.columns[2]

('a', 'b', 'c')

In [132]:
line=pd.DataFrame({df.columns[0]:'--',df.columns[1]:'--',df.columns[2]:'--'},index=[1])

In [133]:
line

a   b   c
1  --  --  --

In [134]:
df0=pd.concat([df.loc[:0],line,df.loc[1:]])
df0

a   b   c
0   1   a   A
1  --  --  --
1   2   b   B
2   3   c   C

In [135]:
#对df0重新索引，方法和1
df1=df0.reset_index()
df1

index   a   b   c
0      0   1   a   A
1      1  --  --  --
2      1   2   b   B
3      2   3   c   C

In [136]:
df2=df1.drop('index',axis=1)

In [137]:
df2

a   b   c
0   1   a   A
1  --  --  --
2   2   b   B
3   3   c   C

In [138]:
#对df0重新索引，方法2。直接对reset_index()函数加drop=True参数。
df2=pd.concat([df.loc[:0],line,df.loc[1:]]).reset_index(drop=True)
df2

a   b   c
0   1   a   A
1  --  --  --
2   2   b   B
3   3   c   C

In [139]:
#对df0重新索引，方法3。先找出df0的索引长度：lenth=len(df0.index),再利用整数序列函数生成索引：range(lenth)，然后把生成的索引赋值给df0.index.
df0.index=range(len(df0.index))
df0

a   b   c
0   1   a   A
1  --  --  --
2   2   b   B
3   3   c   C

## 4.修改记录 

In [90]:
#整体替换
df0['a']=[9,8,7,6]
df0

a   b   c
0  9   a   A
1  8  --  --
2  7   b   B
3  6   c   C

In [96]:
#个别修改
import  pandas as pd
df=pd.read_excel('data/i_nuc.xls',sheet_name='Sheet3')
df

学号        班级   姓名 性别  英语  体育  军训  数分  高代  解几
0   2308024241  23080242   成龙  男  76  78  77  40  23  60
1   2308024244  23080242   周怡  女  66  91  75  47  47  44
2   2308024251  23080242   张波  男  85  81  75  45  45  60
3   2308024249  23080242   朱浩  男  65  50  80  72  62  71
4   2308024219  23080242   封印  女  73  88  92  61  47  46
5   2308024201  23080242   迟培  男  60  50  89  71  76  71
6   2308024347  23080243   李华  女  67  61  84  61  65  78
7   2308024307  23080243   陈田  男  76  79  86  69  40  69
8   2308024326  23080243   余皓  男  66  67  85  65  61  71
9   2308024320  23080243   李嘉  女  62  作弊  90  60  67  77
10  2308024342  23080243  李上初  男  76  90  84  60  66  60
11  2308024310  23080243   郭窦  女  79  67  84  64  64  79
12  2308024435  23080244  姜毅涛  男  77  71  缺考  61  73  76
13  2308024432  23080244   赵宇  男  74  74  88  68  70  71
14  2308024446  23080244   周路  女  76  80  77  61  74  80
15  2308024421  23080244  林建祥  男  72  72  81  63  90  75
16  2308024433  23080244  李大强  男  79  76  77  78  70  70
17  2308024428  23080244  李侧通  男  64  96  91  69  60  77
18  2308024402  23080244   王慧  女  73  74  93  70  71  75
19  2308024422  23080244  李晓亮  男  85  60  85  72  72  83
20  2308024201  23080242   迟培  男  60  50  89  71  76  71

In [97]:
df.replace('作弊',0)

学号        班级   姓名 性别  英语  体育  军训  数分  高代  解几
0   2308024241  23080242   成龙  男  76  78  77  40  23  60
1   2308024244  23080242   周怡  女  66  91  75  47  47  44
2   2308024251  23080242   张波  男  85  81  75  45  45  60
3   2308024249  23080242   朱浩  男  65  50  80  72  62  71
4   2308024219  23080242   封印  女  73  88  92  61  47  46
5   2308024201  23080242   迟培  男  60  50  89  71  76  71
6   2308024347  23080243   李华  女  67  61  84  61  65  78
7   2308024307  23080243   陈田  男  76  79  86  69  40  69
8   2308024326  23080243   余皓  男  66  67  85  65  61  71
9   2308024320  23080243   李嘉  女  62   0  90  60  67  77
10  2308024342  23080243  李上初  男  76  90  84  60  66  60
11  2308024310  23080243   郭窦  女  79  67  84  64  64  79
12  2308024435  23080244  姜毅涛  男  77  71  缺考  61  73  76
13  2308024432  23080244   赵宇  男  74  74  88  68  70  71
14  2308024446  23080244   周路  女  76  80  77  61  74  80
15  2308024421  23080244  林建祥  男  72  72  81  63  90  75
16  2308024433  23080244  李大强  男  79  76  77  78  70  70
17  2308024428  23080244  李侧通  男  64  96  91  69  60  77
18  2308024402  23080244   王慧  女  73  74  93  70  71  75
19  2308024422  23080244  李晓亮  男  85  60  85  72  72  83
20  2308024201  23080242   迟培  男  60  50  89  71  76  71

In [98]:
df.replace({'体育':'作弊'},0)

学号        班级   姓名 性别  英语  体育  军训  数分  高代  解几
0   2308024241  23080242   成龙  男  76  78  77  40  23  60
1   2308024244  23080242   周怡  女  66  91  75  47  47  44
2   2308024251  23080242   张波  男  85  81  75  45  45  60
3   2308024249  23080242   朱浩  男  65  50  80  72  62  71
4   2308024219  23080242   封印  女  73  88  92  61  47  46
5   2308024201  23080242   迟培  男  60  50  89  71  76  71
6   2308024347  23080243   李华  女  67  61  84  61  65  78
7   2308024307  23080243   陈田  男  76  79  86  69  40  69
8   2308024326  23080243   余皓  男  66  67  85  65  61  71
9   2308024320  23080243   李嘉  女  62   0  90  60  67  77
10  2308024342  23080243  李上初  男  76  90  84  60  66  60
11  2308024310  23080243   郭窦  女  79  67  84  64  64  79
12  2308024435  23080244  姜毅涛  男  77  71  缺考  61  73  76
13  2308024432  23080244   赵宇  男  74  74  88  68  70  71
14  2308024446  23080244   周路  女  76  80  77  61  74  80
15  2308024421  23080244  林建祥  男  72  72  81  63  90  75
16  2308024433  23080244  李大强  男  79  76  77  78  70  70
17  2308024428  23080244  李侧通  男  64  96  91  69  60  77
18  2308024402  23080244   王慧  女  73  74  93  70  71  75
19  2308024422  23080244  李晓亮  男  85  60  85  72  72  83
20  2308024201  23080242   迟培  男  60  50  89  71  76  71

In [99]:
df.replace({'体育':'作弊','军训':'缺考'},0)

学号        班级   姓名 性别  英语  体育  军训  数分  高代  解几
0   2308024241  23080242   成龙  男  76  78  77  40  23  60
1   2308024244  23080242   周怡  女  66  91  75  47  47  44
2   2308024251  23080242   张波  男  85  81  75  45  45  60
3   2308024249  23080242   朱浩  男  65  50  80  72  62  71
4   2308024219  23080242   封印  女  73  88  92  61  47  46
5   2308024201  23080242   迟培  男  60  50  89  71  76  71
6   2308024347  23080243   李华  女  67  61  84  61  65  78
7   2308024307  23080243   陈田  男  76  79  86  69  40  69
8   2308024326  23080243   余皓  男  66  67  85  65  61  71
9   2308024320  23080243   李嘉  女  62   0  90  60  67  77
10  2308024342  23080243  李上初  男  76  90  84  60  66  60
11  2308024310  23080243   郭窦  女  79  67  84  64  64  79
12  2308024435  23080244  姜毅涛  男  77  71   0  61  73  76
13  2308024432  23080244   赵宇  男  74  74  88  68  70  71
14  2308024446  23080244   周路  女  76  80  77  61  74  80
15  2308024421  23080244  林建祥  男  72  72  81  63  90  75
16  2308024433  23080244  李大强  男  79  76  77  78  70  70
17  2308024428  23080244  李侧通  男  64  96  91  69  60  77
18  2308024402  23080244   王慧  女  73  74  93  70  71  75
19  2308024422  23080244  李晓亮  男  85  60  85  72  72  83
20  2308024201  23080242   迟培  男  60  50  89  71  76  71

In [100]:
#多值替换
df.replace(['成龙','周怡'],['陈龙','周毅'])

学号        班级   姓名 性别  英语  体育  军训  数分  高代  解几
0   2308024241  23080242   陈龙  男  76  78  77  40  23  60
1   2308024244  23080242   周毅  女  66  91  75  47  47  44
2   2308024251  23080242   张波  男  85  81  75  45  45  60
3   2308024249  23080242   朱浩  男  65  50  80  72  62  71
4   2308024219  23080242   封印  女  73  88  92  61  47  46
5   2308024201  23080242   迟培  男  60  50  89  71  76  71
6   2308024347  23080243   李华  女  67  61  84  61  65  78
7   2308024307  23080243   陈田  男  76  79  86  69  40  69
8   2308024326  23080243   余皓  男  66  67  85  65  61  71
9   2308024320  23080243   李嘉  女  62  作弊  90  60  67  77
10  2308024342  23080243  李上初  男  76  90  84  60  66  60
11  2308024310  23080243   郭窦  女  79  67  84  64  64  79
12  2308024435  23080244  姜毅涛  男  77  71  缺考  61  73  76
13  2308024432  23080244   赵宇  男  74  74  88  68  70  71
14  2308024446  23080244   周路  女  76  80  77  61  74  80
15  2308024421  23080244  林建祥  男  72  72  81  63  90  75
16  2308024433  23080244  李大强  男  79  76  77  78  70  70
17  2308024428  23080244  李侧通  男  64  96  91  69  60  77
18  2308024402  23080244   王慧  女  73  74  93  70  71  75
19  2308024422  23080244  李晓亮  男  85  60  85  72  72  83
20  2308024201  23080242   迟培  男  60  50  89  71  76  71

In [101]:
df.replace({'成龙':'陈龙','周怡':'周毅'}) #或者用df.replace({'成龙','周怡'},{'陈龙','周毅'})

学号        班级   姓名 性别  英语  体育  军训  数分  高代  解几
0   2308024241  23080242   陈龙  男  76  78  77  40  23  60
1   2308024244  23080242   周毅  女  66  91  75  47  47  44
2   2308024251  23080242   张波  男  85  81  75  45  45  60
3   2308024249  23080242   朱浩  男  65  50  80  72  62  71
4   2308024219  23080242   封印  女  73  88  92  61  47  46
5   2308024201  23080242   迟培  男  60  50  89  71  76  71
6   2308024347  23080243   李华  女  67  61  84  61  65  78
7   2308024307  23080243   陈田  男  76  79  86  69  40  69
8   2308024326  23080243   余皓  男  66  67  85  65  61  71
9   2308024320  23080243   李嘉  女  62  作弊  90  60  67  77
10  2308024342  23080243  李上初  男  76  90  84  60  66  60
11  2308024310  23080243   郭窦  女  79  67  84  64  64  79
12  2308024435  23080244  姜毅涛  男  77  71  缺考  61  73  76
13  2308024432  23080244   赵宇  男  74  74  88  68  70  71
14  2308024446  23080244   周路  女  76  80  77  61  74  80
15  2308024421  23080244  林建祥  男  72  72  81  63  90  75
16  2308024433  23080244  李大强  男  79  76  77  78  70  70
17  2308024428  23080244  李侧通  男  64  96  91  69  60  77
18  2308024402  23080244   王慧  女  73  74  93  70  71  75
19  2308024422  23080244  李晓亮  男  85  60  85  72  72  83
20  2308024201  23080242   迟培  男  60  50  89  71  76  71

## 5.交换行和列 

In [141]:
df=pd.DataFrame({'a':[1,2,3],'b':['a','b','c'],'c':['A','B','C']})
df

a  b  c
0  1  a  A
1  2  b  B
2  3  c  C

In [142]:
hang=[0,2,1]
df.reindex(hang)

a  b  c
0  1  a  A
2  3  c  C
1  2  b  B

In [143]:
lie=['a','c','b']
df.reindex(columns=lie)

a  c  b
0  1  A  a
1  2  B  b
2  3  C  c

In [146]:
df.loc[[0,2],:]=df.loc[[2,0],:].values
df

a  b  c
0  3  c  C
1  2  b  B
2  1  a  A

In [120]:
df.loc[:,['c','a']]=df.loc[:,['a','c']] .values #交换两列
df

a  b  c
0  C  c  3
1  B  b  2
2  A  a  1

In [136]:
name=list(df.columns)  #返回['a','b','c']
i=name.index('a')   #值为0
j=name.index('b')
name

['a', 'b', 'c']

In [137]:
name[i],name[j]=name[j],name[i]
df.columns=name
df

b  a  c
0  C  c  3
1  B  b  2
2  A  a  1

In [138]:
name

['b', 'a', 'c']

#### *增加一列

In [147]:
import pandas as pd
df=pd.DataFrame({'a':[1,2,3],'b':['a','b','c'],'c':['A','B','C']})
line=pd.DataFrame({df.columns[0]:'--',df.columns[1]:'--',df.columns[2]:'--'},index=[1])
df0=pd.concat([df.loc[:0],line,df.loc[1:]])
df0.index=range(len(df0.index))
df0

a   b   c
0   1   a   A
1  --  --  --
2   2   b   B
3   3   c   C

In [148]:
df0['d']=range(len(df0.index))
df0

a   b   c  d
0   1   a   A  0
1  --  --  --  1
2   2   b   B  2
3   3   c   C  3

In [149]:
df0.loc[:,['b','d']]=df0.loc[:,['d','b']].values
df0

a  b   c   d
0   1  0   A   a
1  --  1  --  --
2   2  2   B   b
3   3  3   C   c

In [150]:
lie=list(df0.columns)
i=lie.index('b')
j=lie.index('d')
lie[i],lie[j]=lie[j],lie[i]
df0.columns=lie
df0

a  d   c   b
0   1  0   A   a
1  --  1  --  --
2   2  2   B   b
3   3  3   C   c